src/java/org/apache/cassandra/index/sai/memory/TrieMemtableIndex.java

```diff
@@ -256,38 +296,105 @@ public long estimateMatchingRowsCount(Expression expression, AbstractBounds<Part
     {
         if (keys.isEmpty())
             return CloseableIterator.emptyIterator();
-        return SortingIterator.createCloseable(
-            orderer.getComparator(),
-            keys,
-            key ->
+
+        if (!orderer.isBM25())
+        {
+            return SortingIterator.createCloseable(
+                orderer.getComparator(),
+                keys,
+                key ->
+                {
+                    var partition = memtable.getPartition(key.partitionKey());
+                    if (partition == null)
+                        return null;
+                    var row = partition.getRow(key.clustering());
+                    if (row == null)
+                        return null;
+                    var cell = row.getCell(indexContext.getDefinition());
+                    if (cell == null)
+                        return null;
+
+                    // We do two kinds of encoding... it'd be great to make this more straight forward, but this is what
+                    // we have for now. I leave it to the reader to inspect the two methods to see the nuanced differences.
+                    var encoding = encode(TypeUtil.encode(cell.buffer(), validator));
+                    return new PrimaryKeyWithByteComparable(indexContext, memtable, key, encoding);
+                },
+                Runnables.doNothing()
+            );
+        }
+
+        // BM25
+        var analyzer = indexContext.getAnalyzerFactory().create();
+        var queryTerms = orderer.getQueryTerms();
+        var docStats = computeDocumentFrequencies(queryContext, queryTerms);
+        var it = keys.stream().map(pk -> BM25Utils.DocTF.createFromDocument(pk, getCellForKey(pk), analyzer, queryTerms)).iterator();
+        return BM25Utils.computeScores(CloseableIterator.wrap(it),
+                                       queryTerms,
+                                       docStats,
+                                       indexContext,
+                                       memtable);
+    }
+
+    /**
+     * Count document frequencies for each term using brute force
+     */
+    private BM25Utils.DocStats computeDocumentFrequencies(QueryContext queryContext, List<ByteBuffer> queryTerms)
+    {
+        var termIterators = keyIteratorsPerTerm(queryContext, Bounds.unbounded(indexContext.getPartitioner()), queryTerms);
+        var documentFrequencies = new HashMap<ByteBuffer, Long>();
+        for (int i = 0; i < queryTerms.size(); i++)
+        {
+            // KeyRangeIterator.getMaxKeys is not accurate enough, we have to count them
+            long keys = 0;
+            for (var it = termIterators.get(i); it.hasNext(); it.next())
+                keys++;
+            documentFrequencies.put(queryTerms.get(i), keys);
+        }
+        long docCount = 0;
+
+        // count all documents in the queried column
```

Comment:
Did you consider keeping track of this at write time so that we don't need to recompute it for each read? This seems like metadata that each MemoryIndex could keep track of with a `LongAdder`.

Reply:
yes, but since it's in memory I decided to leave it as "optimize this if it turns out to be a hot spot"